Added parameters for watertight sideset and nodeset diagnostic check

DanielYankura opened the pull request:
-Diagnostics can now take an additional vector parameter _watertight_boundaries 
-Users can use it to input a list of sidesets/nodesets 
-If supplied the diagnostic check functionality changes 
-It will instead check if an external side/node is assigned to one of the provided boundary ids 
-The error message for these instances is worded differently to reflect what the diagnostic check is looking for 
-The t in checkWaterTightSidesets is now lowercase because it was bothering me and to match checkWatertightNodesets

closes #29077

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->
To provide users with more mesh debugging tools

## Design
<!--A concise description (design) of the enhancement.-->
Users  can supply a list of boundary ids. If supplied this diagnostic check looks for sides/nodes that are external and that are not assigned to one of the prescribed boundaries.
## Impact
<!--Will the enhancement change existing APIs or add something new?-->
This modifies the checkWatertightSideset and checkWatertightNodeset diagnostic tools. If no list of boundary ids is supplied however, then they behave the same as they did before.

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

DanielYankura committed:
Added parameters for watertight sideset and nodeset diagnostic check

-Diagnostics can now take an additional vector parameter _watertight_boundaries
-Users can use it to input a list of sidesets/nodesets
-If supplied the diagnostic check functionality changes
-It will instead check if an external side/node is assigned to one of the provided boundary ids
-The error message for these instances is worded differently to reflect what the diagnostic check is looking for
-The t in checkWaterTightSidesets is now lowercase because it was bothering me and to match checkWatertightNodesets

closes #29077

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2562698/), step Clang format on 3c505cd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29136/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29136/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9446d5fc904cc635205b563219f91889cd5a678c`

DanielYankura committed:
clang formart changes

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2562805/), step Docs: sync website on 41d5897 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29136/site)

This comment will be updated on new commits.

GiudGiud commented on framework/include/meshgenerators/MeshDiagnosticsGenerator.h:
```diff
@@ -32,9 +32,13 @@ class MeshDiagnosticsGenerator : public MeshGenerator
   void checkWatertightNodesets(const std::unique_ptr<MeshBase> & mesh) const;
+  /// Helper function that finds the intersection of the _watertight_boundaries vector and the provided one
+  std::vector<boundary_id_type>
+  findBoundaryOverlap(std::vector<boundary_id_type>,
+                      std::vector<boundary_id_type> boundary_ids) const;
```
```suggestion
  findBoundaryOverlap(const std::vector<boundary_id_type> & boundary_copy,
                      const std::vector<boundary_id_type> & boundary_ids) const;
```

DanielYankura replied:
So in this function I do sort the one vector. If I do that then I don't think I can make it const?

roystgnr replied:
No, you can't.  You should put the requirement for the vector in a comment above the method, though, and add an assertion for it in the method.

Don't both vectors have to be sorted for a `set_intersection()`, though?

DanielYankura replied:
Yes they do. The one vector (boundary_copy) is a copy of the const vector that contains the list of boundaries specified in the input file. It's sorted before it's inputted. The other is sorted inside the function. I feel like I had a good reason for sorting the one outside of the function and one inside but I can't remember why. If it comes to me I'll let you know,

roystgnr replied:
It's reasonable to require a vector to be sorted outside the function for performance reasons (you can then reuse the sorted property over multiple invocations with the same vector with no call to a sort inside the function, which is good since even a no-op sort is still O(N)).  It's a little weird to require this of one vector but not the other, but for a specialized little function like this and a use case where you only reuse one input that is the simplest thing to do.

Why not pass the unsorted vector by non-const reference, though?  If you have a non-const vector to hand it, and you're either not reusing or not relying on the ordering of that vector afterwards, which seems to be the case here, then it's faster to sort the input rather than copying the input and sorting that.

GiudGiud commented on framework/include/meshgenerators/MeshDiagnosticsGenerator.h:
```diff
@@ -32,9 +32,13 @@ class MeshDiagnosticsGenerator : public MeshGenerator
+  void checkWatertightSidesets(const std::unique_ptr<MeshBase> & mesh) const;
   //// Routine to check is mesh is fully covered in nodesets
   void checkWatertightNodesets(const std::unique_ptr<MeshBase> & mesh) const;
+  /// Helper function that finds the intersection of the _watertight_boundaries vector and the provided one
```
it does not use _waterright_boundaries, but rather a copy that is sorted.
so just write the description in a more general manner that does not mention _watertight_boundaries

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -49,6 +49,8 @@ MeshDiagnosticsGenerator::validParams()
       "check_for_watertight_nodesets",
       chk_option,
       "whether to check for external nodes that are not assigned to any nodeset");
+  params.addParam<std::vector<BoundaryID>>(
```
```suggestion
  params.addParam<std::vector<BoundaryName>>(
```
let s do names. ids, especially for boundaries, are really error-prone

DanielYankura replied:
So I was looking into this and I think I ran into a problem. For boundary ids there's a map from element to boundary id but there's no equivalent for element to boundary names. There are functions for getting boundary names from the ID but I feel like that defeats the purpose?

GiudGiud replied:
the user interface asks for names.

then shortly after in the constructor or the setup we take the vector of names and convert it to a vector of ids

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -49,6 +49,8 @@ MeshDiagnosticsGenerator::validParams()
       chk_option,
       "whether to check for external nodes that are not assigned to any nodeset");
+  params.addParam<std::vector<BoundaryID>>(
+      "boundaries_to_check", {}, "IDs of boundaries to be checked for watertight checks");
```
```suggestion
      "boundaries_to_check", {}, "Names boundaries that should form a watertight envelope around the mesh. Defaults to all the boundaries combined.");
```

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -300,21 +303,28 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
   const auto sideset_map = boundary_info.get_sideset_map();
   unsigned int num_faces_without_sideset = 0;
 
+  // Make copy of _watertight_boundaries and sort it for later use
+  std::vector<BoundaryID> boundary_copy = _watertight_boundaries;
+  std::sort(boundary_copy.begin(), boundary_copy.end());
```
would not hurt to just store it sorted

DanielYankura replied:
So is there a reason that the vector of boundary IDs needs to be constant? I was thinking that if it wasn't constant that I could sort it immediately after it's read in and that way it could be stored sorted. Do you think that would cause any issues down the road?

roystgnr replied:
I'd leave it non-const if that's the cost to store it sorted without requiring some ugly function in the initializer list.  Everything you can mark `const` is one more clue to future programmers and one less thing they can accidentally break later, so all else being equal I love `const`, but here all else is not all equal.

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -300,21 +303,28 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
-        // If external check if that side had a sideset
+        // If external get the boundary ids associated with this side
+        std::vector<boundary_id_type> boundary_ids;
+        // boundary_info.boundary_ids(elem, i, boundary_ids);
```
```suggestion
```

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -330,6 +340,25 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
             num_faces_without_sideset++;
           }
         }
+        else if (!_watertight_boundaries.empty() && intersections.empty())
```
can you consolidate all of this code by doing:
if (...... OR (!_watertight_boundaries.empty() && intersections.empty()))

DanielYankura replied:
I separated the two because I figured that was the best way to have different error messages depending on whether it just has no sideset, or was assigned to the wrong one? If we want them combined I could make the error message more generic?

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -391,6 +429,21 @@ MeshDiagnosticsGenerator::checkWatertightNodesets(const std::unique_ptr<MeshBase
+              _console << message << std::endl;
+            }
+          }
         }
```
same comments here

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -401,6 +454,22 @@ MeshDiagnosticsGenerator::checkWatertightNodesets(const std::unique_ptr<MeshBase
 }
 
+std::vector<boundary_id_type>
+MeshDiagnosticsGenerator::findBoundaryOverlap(std::vector<boundary_id_type> boundary_copy,
+                                              std::vector<boundary_id_type> boundary_ids) const
```
```suggestion
MeshDiagnosticsGenerator::findBoundaryOverlap(const std::vector<boundary_id_type> & boundary_copy,
                                              std::vector<boundary_id_type> & boundary_ids) const
```

DanielYankura committed:
made various changes for PR

-Changed wording of comments to be more accurate
-_watertight_boundaries is no longer const and is sorted before being called
-findBoundaryOverlap changed to pass by reference
-Still need to change BoundaryID to BoundaryName before merging

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2579331/), step Clang format on 86878b6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29136/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29136/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 507f39e9e10c7b6801458841cac71acbfa8731c9`

DanielYankura committed:
style changes

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2579444/), step Generate coverage on d8ae30e wanted to post the following:


## Framework coverage

Inconsistent report tags were found between the head and base reports.
This can happen when reports are missing from either the head or the base.

Inconsistent tags:
  framework-torch-gpu
[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29136/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29136/coverage/xfem)

</details>



This comment will be updated on new commits.

DanielYankura committed:
consolidated error message code

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -303,33 +312,38 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
+        std::string message;
+        if (mesh->mesh_dimension() == 3)
+          message = "Element " + std::to_string(elem->id()) +
+                    " contains an external face which has not been assigned to "; // a sideset";
+        else
+          message = "Element " + std::to_string(elem->id()) +
+                    " contains an external edge which has not been assigned to "; // a sideset";
```
```suggestion
                    " contains an external face which has not been assigned to ";
        else
          message = "Element " + std::to_string(elem->id()) +
                    " contains an external edge which has not been assigned to ";
```

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -303,33 +312,38 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
+        if (no_specified_ids)
+          message = message + "a sideset";
+        else if (specified_ids)
+          message = message + "one of the specifiec sidesets";
```
```suggestion
          message = message + "one of the specified sidesets";
```

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -377,20 +395,26 @@ MeshDiagnosticsGenerator::checkWatertightNodesets(const std::unique_ptr<MeshBase
+          if (no_specified_ids)
+            message = message + "a nodeset";
+          else if (specified_ids)
+            message = message + "one of the specifiec nodesets";
```
```suggestion
            message = message + "one of the specified nodesets";
```

GiudGiud reviewed (commented):
almost ready

DanielYankura committed:
BoundaryID to Boundary Name

- watertight boundary params is now a vector of BoundaryName
- User inputs boundary names and they are then converted to IDs
- spelling mistakes corrected

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -137,11 +144,15 @@ MeshDiagnosticsGenerator::generate()
   mesh->prepare_for_use();
 
+  // convert BoundaryNames to IDs and sort
+  _watertight_boundaries = prepareBoundaries(mesh);
+  //_watertight_boundaries = mesh->getBoundaryIDs(_watertight_boundary_names);
```
```suggestion
  _watertight_boundaries = mesh->getBoundaryIDs(_watertight_boundary_names);
```
the first one has a general name but only applies to this one task in this one diagnostics

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -173,6 +184,22 @@ MeshDiagnosticsGenerator::generate()
+  {
+    for (const auto & name : _watertight_boundary_names)
+    {
+      BoundaryID id = MooseMeshUtils::getBoundaryID(name, *mesh);
```
you can use  MooseMeshUtils::getBoundaryIDs(...., *mesh); in the constructor maybe?

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -312,36 +339,42 @@ MeshDiagnosticsGenerator::checkWaterTightSidesets(const std::unique_ptr<MeshBase
   boundary_info.build_side_list();
   const auto sideset_map = boundary_info.get_sideset_map();
   unsigned int num_faces_without_sideset = 0;
+  //
```
```suggestion
```

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -374,6 +407,10 @@ MeshDiagnosticsGenerator::checkWatertightNodesets(const std::unique_ptr<MeshBase
 
+  // Make copy of _watertight_boundaries and sort it for later use
+  std::vector<BoundaryID> boundary_copy = _watertight_boundaries;
+  std::sort(boundary_copy.begin(), boundary_copy.end());
```
do we need the copy? why not sort the original?

DanielYankura committed:
various changes for PR

- Simplified method for converting boundaryNames to IDs
- Added check to make sure specified boundaries_to_check are all valid
- removed copy of vector that wasn't needed

GiudGiud commented:
Test failures (ARM python thing, fsi timeout, diff in NS) are unrelated
